Source/CombatExtended/CombatExtended/Comps/CompVariableAmmoUser.cs

```diff
@@ -0,0 +1,167 @@
+        {
+            base.PostExposeData();
+            Scribe_Defs.Look(ref SelectedAmmoSet, "SelectedAmmoSet");
+        }
+
+        public override void Initialize(CompProperties vprops)
+        {
+            base.Initialize(vprops);
+            SelectedAmmoSet = Props.ammoSet;
+            RegenSelectedAmmo();
+        }
+        public void RegenSelectedAmmo()
+        {
+            if (currentAmmoInt == null)
+            {
+                currentAmmoInt = (AmmoDef)CurAmmoSet.ammoTypes[0].ammo;
+            }
+            if (selectedAmmo == null)
+            {
+                selectedAmmo = currentAmmoInt;
+            }
+            else
+            {
+                //turned out it was bold to assume similar ammosets always have same ammo types
+                var tempAmmoLink = CurAmmoSet.ammoTypes.Where(l => l.ammo.ammoClass == currentAmmoInt.ammoClass);
+                if (tempAmmoLink.Any())
+                {
+                    selectedAmmo = tempAmmoLink.First().ammo;
+                }
+                else
+                {
+                    selectedAmmo = CurAmmoSet.ammoTypes.First().ammo;
+                }
+            }
+        }
+
+        [Compatibility.Multiplayer.SyncMethod]
+        private void SyncedSelectAmmoSet(AmmoSetDef caliber)
+        {
+            SelectedAmmoSet = caliber;
+            RegenSelectedAmmo();
+        }
+
+        public override IEnumerable<Gizmo> CompGetGizmosExtra()
+        {
+            foreach (var gizmo in base.CompGetGizmosExtra())
+            {
+                yield return gizmo;
+            }
+            if (!Controller.settings.GenericAmmo)
+            {
+                Command_Action command_Action = new Command_Action();
+                command_Action.defaultLabel = "CE_SelectAmmoSet".Translate();
+                command_Action.defaultDesc = "CommandSelectMineralToScanForDesc".Translate();
+                command_Action.icon = ContentFinder<Texture2D>.Get("UI/Buttons/Reload", reportFailure: false);
+                command_Action.action = delegate
+                {
+                    List<FloatMenuOption> list = new List<FloatMenuOption>();
+                    foreach (AmmoSetDef caliber in UsableAmmoSets)
+                    {
+                        FloatMenuOption item = new FloatMenuOption(
+                           caliber.LabelCap,
+                           delegate { SyncedSelectAmmoSet(caliber); },
+                           caliber.ammoTypes.First().ammo,
+                           priority: MenuOptionPriority.Default,
+                           mouseoverGuiAction: delegate (Rect rect) { ContainedAmmoPopOut(rect, caliber); });
+                        list.Add(item);
+                    }
+                    Find.WindowStack.Add(new FloatMenu(list));
+                };
+                yield return command_Action;
+            }
+        }
+
+        AmmoSetDef currentlyHoveredOverAmmoSet = null;
```

Comment:
Be explicit about things being private, rather than using default visibility.

Also, is there a reason to stick these fields down here, instead of at the top with the rest of the fields?

Reply:
changes done.